core/src/Cabana_Experimental_NeighborList.hpp

```diff
@@ -238,40 +258,63 @@ template <typename DeviceType, typename Slice, typename Tag>
 auto make2DNeighborList( Tag, Slice const &coordinate_slice,
                          typename Slice::size_type first,
                          typename Slice::size_type last,
-                         typename Slice::value_type radius )
+                         typename Slice::value_type radius,
+                         int buffer_size = 0 )
 {
+    assert( buffer_size >= 0 );
+
     using MemorySpace = typename DeviceType::memory_space;
     using ExecutionSpace = typename DeviceType::execution_space;
     ExecutionSpace space{};
 
     ArborX::BVH<MemorySpace> bvh( space, coordinate_slice );
 
-    Kokkos::View<int *, DeviceType> indices( "indices", 0 );
-    Kokkos::View<int *, DeviceType> offset( "offset", 0 );
-
     auto const predicates =
         Impl::makePredicates( coordinate_slice, first, last, radius );
 
     auto const n_queries =
         ArborX::AccessTraits<decltype( predicates ),
                              ArborX::PredicatesTag>::size( predicates );
 
+    Kokkos::View<int **, DeviceType> neighbors;
     Kokkos::View<int *, DeviceType> counts( "counts", n_queries );
-    bvh.query(
-        space, predicates,
-        Impl::NeighborDiscriminatorCallback2D_FirstPass<decltype( counts ),
-                                                        Tag>{counts},
-        indices, offset );
+    if ( buffer_size > 0 )
+    {
+        neighbors = Kokkos::View<int **, DeviceType>(
+            Kokkos::view_alloc( "neighbors", Kokkos::WithoutInitializing ),
+            n_queries, buffer_size );
+        bvh.query(
+            space, predicates,
+            Impl::NeighborDiscriminatorCallback2D_FirstPass_BufferOptimization<
+                decltype( counts ), decltype( neighbors ), Tag>{counts,
+                                                                neighbors} );
+    }
+    else
+    {
+        bvh.query(
+            space, predicates,
+            Impl::NeighborDiscriminatorCallback2D_FirstPass<decltype( counts ),
+                                                            Tag>{counts} );
+    }
 
-    Kokkos::View<int **, DeviceType> neighbors(
+    auto const max_neighbors = ArborX::max( space, counts );
+    if ( max_neighbors <= buffer_size )
+    {
+        // NOTE We do not bother shrinking to eliminate the excess allocation.
+        // NOTE If buffer_size is 0, neighbors is default constructed.  This is
+        // fine with the current design/implementation of NeighborList access
+        // traits.
+        return Dense<MemorySpace, Tag>{counts, neighbors, first, bvh.size()};
+    }
+
+    neighbors = Kokkos::View<int **, DeviceType>(
         Kokkos::view_alloc( "neighbors", Kokkos::WithoutInitializing ),
```

Comment:
Should we start using `view_alloc` throughout?

Reply:
Up to you.  At this point I am not sure we will actually ever go further than kokkos/kokkos#3264 and actually pull the trigger on deprecating `ViewAllocateWithoutInitializing`.